snorkel/candidates.py

```diff
@@ -101,80 +155,39 @@ def _extract_multiprocess(self, candidate_space, matcher, contexts, parallelism=
 
-    def __iter__(self):
-        """Default iterator is over Candidates"""
-        return self._candidates_by_id.itervalues()
+class Ngram(Candidate):
+    """
+    A span of _n_ tokens, identified by Context id and character-index start, end (inclusive).
 
-    def get_candidates(self):
-        return self._candidates_by_id.values()
+    char_offsets are **relative to the Document start**
+    """
+    __tablename__ = 'ngram'
+    id = Column(String, ForeignKey('candidate.id'), primary_key=True)
+    char_start = Column(Integer)
+    char_end = Column(Integer)
+    meta = Column(PickleType)
 
-    def get_candidate(self, id):
-        """Retrieve a candidate by candidate id"""
-        return self._candidates_by_id[id]
-    
-    def get_candidates_in(self, context_id):
-        """Return the candidates in a specific context (e.g. Sentence)"""
-        return self._candidates_by_context_id[context_id]
+    __mapper_args__ = {
+        'polymorphic_identity': 'ngram',
+    }
 
-    def gold_stats(self, gold_set):
-        """Return precision and recall relative to a "gold" set of candidates of the same type"""
-        gold = gold_set if isinstance(gold_set, set) else set(gold_set)
-        cs   = self.get_candidates()
-        nc   = len(cs)
-        ng   = len(gold)
-        both = len(gold.intersection(cs))
-        print "# of gold annotations\t= %s" % ng
-        print "# of candidates\t\t= %s" % nc
-        print "Candidate recall\t= %0.3f" % (both / float(ng),)
-        print "Candidate precision\t= %0.3f" % (both / float(nc),)
-    
+    def __len__(self):
+        return self.char_end - self.char_start + 1
 
-# Basic sentence attributes
-WORDS        = 'words'
-CHAR_OFFSETS = 'char_offsets'
-TEXT         = 'text'
+    def get_word_start(self):
+        return self.char_to_word_index(self.char_start)
 
-class Ngram(Candidate):
-    """A span of _n_ tokens, identified by sentence id and character-index start, end (inclusive)"""
-    def __init__(self, char_start, char_end, sent, metadata={}):
-        
-        # Inherit full sentence object (tranformed to dict) and check for necessary attribs
-        self.sentence = sent if isinstance(sent, dict) else sent._asdict()
-        self.sent_id  = self.sentence['id']
-        REQ_ATTRIBS = ['id', WORDS]
-        if not all([self.sentence.has_key(a) for a in REQ_ATTRIBS]):
-            raise ValueError("Sentence object must have attributes %s to form Ngram object" % ", ".join(REQ_ATTRIBS))
-
-        # Set basic object attributes
-        self.id          = "%s:%s-%s" % (self.sent_id, char_start, char_end)
-        self.char_start  = char_start
-        self.char_end    = char_end
-        self.char_len    = char_end - char_start + 1
-        self.word_start  = self.char_to_word_index(char_start)
-        self.word_end    = self.char_to_word_index(char_end)
-        self.n           = self.word_end - self.word_start + 1
-
```

Comment:
Sentence level helper methods added in c9f9c7c

metadata has been renamed to meta because metadata is a reserved SQLAlchemy name.